### PR TITLE
[ShapeOpt] Feature MinMax or MaxMin optimization

### PR DESCRIPTION
**Pull request by dbaumgaertner**
This PR adds a simple feature to combine objective functions using a min- or a max-function. Hence it is possible to realize a MinMax or MaxMin optimization. 

The feature is compatible and combinable with the already existing weighted sum approach. Also,  objective functions may be combined in a nested hierarchy. 

A unit test is added to demonstrate and test the feature.

### CODE REVIEW

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py [hunk 1]
```diff
@@ -36,11 +36,11 @@ def CreateAnalyzer(optimization_settings, model_part_controller, external_analyz
         internal_analyzer = EmptyAnalyzer()
 
     dependency_graph, exist_dependencies = _CreateDependencyGraphRecursively(objectives)
-    # dependency_graph = [ (response_id_1, [], weight_1),
-    #                      (response_id_2, [], weight_2),
-    #                      (response_id_3, [ (response_id_3a, [], weight_3a),
-    #                                        (response_id_3b, [], weight_3b),
-    #                                        (response_id_3c, [...], weight_3c) ], weight_3),
+    # dependency_graph = [ (response_id_1, weight_1, []),
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py]: dependency type is missing in this description of the tree

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py [hunk 2]
```diff
@@ -193,57 +194,82 @@ def __RequestResponsesAccordingDependencies(self, communicator):
...
+            if combined_value is None:
+                combined_value = value
+                relevant_responses.append(response_id)
+            else:
+                if combination_type == "sum":
+                    combined_value += value
+                    relevant_responses.append(response_id)
+                elif combination_type == "max":
+                    if value > combined_value:
+                        relevant_responses[0] = response_id
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py]:
```python 
relevant_responses = [response_id]
```
makes the intent even clearer (IMO)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py [hunk 3]
```diff
@@ -193,57 +194,82 @@ def __RequestResponsesAccordingDependencies(self, communicator):
...
+                        combined_value = value
+                elif combination_type == "min":
+                    if value < combined_value:
+                        relevant_responses[0] = response_id
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py]: same

File: applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py [hunk 4]
```diff
@@ -38,6 +38,13 @@ def initializeCommunication(self):
         self.__deleteAllRequests()
         self.__deleteAllReportedValues()
 
+    # --------------------------------------------------------------------------
+    def initializeRequest(self, response_id, value_request=True, gradient_request=True):
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py]: this function name does not follow the convention, but it seems the whole file is like this, so up to you if you want to change it here.

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py [hunk 5]
```diff
@@ -193,57 +194,82 @@ def __RequestResponsesAccordingDependencies(self, communicator):
...
+                        combined_value = value
+                else:
+                    raise NameError("The following combination_type is not supported:" + combination_type)
+
+        for response_id, _, _, _ in dependencies:
+            if response_id not in relevant_responses:
+                communicator.initializeRequest(response_id, value_request=False)
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py]:
```suggestion
                communicator.UpdateRequest(response_id, value_request=True, gradient_request=False)
```
To my understanding at this stage you have already used the value, so it should "remain" requested, but you can tell now that the gradient is not needed anymore?
Or are the requests set to false when the corresponding values are set at the communicator?

**Review by armingeiser — CHANGES_REQUESTED** [review]
Thanks, clean as always!
Only real request is about the usage of the new funciton to set the value/gradient requests at the communicator (see comments)

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py (hunk 5 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py]: Anyway i would prefer to write both keyword arguments here

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py]: This shows how kwargs can be enforced: https://stackoverflow.com/questions/2965271/forced-naming-of-parameters-in-python

File: applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py (hunk 4 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py]: I would leave this to another PR, as it affects many files in the code. This should be done in a cleaning step at some point.

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py]: good for me

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py (hunk 5 above)
Comment (dbaumgaertner) [reply applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py]: I adjusted it to use kwargs. Totally forgot about this feature. Thanks for the hint. Also, I used your suggested function name as it expresses the intent much better.

In fact, it is an updating of the requests within the analyzer. Also, just the gradient request ist updated right after combining the values (determining the max value).

**Review by armingeiser — APPROVED** [review]

**Pull request merged**